rust/src/pop3/pop3.rs

```diff
@@ -0,0 +1,767 @@
+                }
+                Ok((rem, None)) => {
+                    // Not enough data. This parser doesn't give us a good indication
+                    // of how much data is missing so just ask for one more byte so the
+                    // parse is called as soon as more data is received.
+
+                    let consumed = input.len() - rem.len();
+                    let needed = rem.len() + 1;
+                    return AppLayerResult::incomplete(consumed as u32, needed as u32);
+                }
+                Err(SawpError {
+                    kind: SawpErrorKind::Incomplete(sawp::error::Needed::Size(needed)),
+                }) => {
+                    let consumed = input.len() - start.len();
+                    let needed = start.len() + needed.get();
+                    return AppLayerResult::incomplete(consumed as u32, needed as u32);
+                }
+                Err(_) => return AppLayerResult::err(),
+            }
+        }
+
+        // Input was fully consumed.
+        return AppLayerResult::ok();
+    }
+
+    fn parse_response(&mut self, input: &[u8], flow: *const Flow) -> AppLayerResult {
+        // We're not interested in empty responses.
+        if input.is_empty() {
+            return AppLayerResult::ok();
+        }
+
+        if self.response_gap {
+            unsafe {
+                if probe(input) != ALPROTO_POP3 {
+                    // The parser now needs to decide what to do as we are not in sync.
+                    // For this pop3, we'll just try again next time.
+                    return AppLayerResult::ok();
+                }
+            }
+
+            // It looks like we're in sync with a message header, clear gap
+            // state and keep parsing.
+            self.response_gap = false;
+        }
+        let mut start = input;
+        while !start.is_empty() {
+            match POP3_PARSER.parse(start, Direction::ToClient) {
+                Ok((rem, Some(msg))) => {
+                    if let InnerMessage::Response(mut response) = msg.inner {
+                        let mut tx = if let Some(tx) = self.find_request() {
+                            tx
+                        } else {
+                            // Server sends greeting before any requests
+                            let tx = self.new_tx();
+                            let tx_id = tx.id();
+                            self.transactions.push_back(tx);
+                            self.get_tx_mut(tx_id - 1).unwrap()
+                        };
+
+                        tx.error_flags |= msg.error_flags;
+
+                        if response.status == sawp_pop3::Status::OK && tx.request.is_some() {
+                            let request = tx.request.as_ref().unwrap();
+                            match &request.keyword {
+                                sawp_pop3::Keyword::STLS => {
+                                    unsafe { AppLayerRequestProtocolTLSUpgrade(flow) };
```

Comment:
would be good to check the retval and set an event if the upgrade was denied. This can happen in certain cases, for example if you would put multiple upgrades in a row. Fuzzing has found some complex chains with multiple protocols as well.

Reply:
Line 215 is checking the retval `response.status == sawp_pop3::Status::OK`, if it was denied this status would be `ERR`